api/v1alpha1/limitador_types.go

```diff
@@ -102,6 +107,93 @@ type LimitadorList struct {
 	Items           []Limitador `json:"items"`
 }
 
+// StorageType defines the valid options for storage
+// +kubebuilder:validation:Enum=memory;redis;redis_cached
+type StorageType string
+
+const (
+	StorageTypeInMemory    StorageType = "memory"
+	StorageTypeRedis                   = "redis"
+	StorageTypeRedisCached             = "redis_cached"
+)
+
+// Storage contains the options for Limitador counters database or in-memory data storage
+type Storage struct {
+	// +optional
+	Redis *Redis `json:"redis,omitempty"`
+
+	// +optional
+	RedisCached *RedisCached `json:"redis-cached,omitempty"`
+}
+
+func (s *Storage) Validate() bool {
+	return s.Redis != nil && s.Redis.ConfigSecretRef != nil ||
+		s.RedisCached != nil && s.RedisCached.ConfigSecretRef != nil
+}
+
+func (s *Storage) SecretRef() *corev1.ObjectReference {
+	if s.Redis != nil {
+		return s.Redis.ConfigSecretRef
+	}
+	return s.RedisCached.ConfigSecretRef
+}
+
+func (s *Storage) Config(url string) []string {
+	if s.Redis != nil {
+		return []string{
+			StorageTypeRedis,
+			url,
+		}
+	} else if s.RedisCached != nil {
+		return []string{
+			StorageTypeRedisCached,
+			url,
+			fmt.Sprintf("--ttl %d", s.RedisCached.Options.TTL),
+			fmt.Sprintf("--ratio %d", s.RedisCached.Options.Ratio),
+			fmt.Sprintf("--flush-period %d", s.RedisCached.Options.FlushPeriod),
+			fmt.Sprintf("--max-cached %d", s.RedisCached.Options.MaxCached),
+		}
+	}
+	return []string{string(StorageTypeInMemory)}
+}
+
+type Redis struct {
+	// +ConfigSecretRef refers to the secret holding the URL for Redis.
+	// +optional
+	ConfigSecretRef *corev1.ObjectReference `json:"configSecretRef,omitempty"`
+}
+
+type RedisCachedOptions struct {
+	// +optional
+	// +kubebuilder:default=5000
+	// TTL for cached counters in milliseconds [default: 5000]
+	TTL int `json:"ttl,omitempty"`
```

Comment:
if the type is `int`, I cannot set TTL to 0. I do not know if that is a use case, but just heads up on this.

Same for all the remaining fields.

Usually I use pointer types for optional fields.

Reply:
0 is included in type `int`, no? Or I'm missing something? I could change it to pointer types too, but I thought there will always have a value because of ` 	// +kubebuilder:default=5000 `

Reply:
While it is true that I can create (with kubectl) a Limitador CR with TTL to 0 and the operator will read it correctly, if I  manage objects using go and golang serialization/deserialization approach, things get weird.

```go
limitadorObj2 := &limitadorv1alpha1.Limitador{                              
    TypeMeta: metav1.TypeMeta{                                              
        Kind:       "Limitador",                                            
        APIVersion: limitadorv1alpha1.GroupVersion.String(),                
    },                                                                      
    ObjectMeta: metav1.ObjectMeta{                                          
        Name:      "limitador2",                                            
        Namespace: limitadorObj.Namespace,                                  
    },                                                                      
    Spec: limitadorv1alpha1.LimitadorSpec{                                  
        Storage: &limitadorv1alpha1.Storage{                                
            RedisCached: &limitadorv1alpha1.RedisCached{                    
                Options: &limitadorv1alpha1.RedisCachedOptions{             
                    TTL:   0,                                               
                    Ratio: 4,                                               
                },                                                          
            },                                                              
        },                                                                  
    },                                                                      
}                                                                           
```

`spec.storage.redis-cached.options.ttl` it set to 0. I create this object

```go
if err := r.CreateResource(ctx, limitadorObj2); err != nil {
    logger.Error(err, "limitadorobj2 not created")          
    return ctrl.Result{}, nil                               
}                                                           
logger.Info("limitadorobj2 created")                        
```

Then I read this object from the cluster:
```go
limitadorObj3 := &limitadorv1alpha1.Limitador{}                                                      
if err := r.GetResource(ctx, client.ObjectKeyFromObject(limitadorObj2), limitadorObj3); err != nil { 
    logger.Error(err, "get resource limitador2")                                                     
    return ctrl.Result{}, nil                                                                        
}          
```

Then, the TTL value will be `5000`. The reason is that when it gets serialized to the apiserver, the TTL field is omitted (the tag `omitempty`). Then, when you request the object to the apiserver, the apiserver will fill not existing values with default values. 

you could remove the `omitempty` tag, but then the field would not be optional.

Reply:
I see, thanks for the explanation! Even if there's no use case for a TTL 0, the expected behaviour won't be sound. I'll address it accordingly. Thanks again!

Reply:
Thing is that the kuadrant operator creates the limitador object from golang code and that ringed a bell